File: backend/app/schemas/notes.py
# ABOUTME: Pydantic schemas for case notes API (Sherlock's Diary).
# ABOUTME: Defines request/response models for note CRUD and export operations.

from datetime import datetime
from enum import Enum
from uuid import UUID

from pydantic import BaseModel, Field


class NoteType(str, Enum):
    """Type of note."""

    TEXT = "TEXT"
    AUDIO = "AUDIO"


class NoteBase(BaseModel):
    """Base schema for note data."""

    type: NoteType
    content: str | None = Field(None, description="Text content for text notes")
    title: str | None = Field(
        None, max_length=255, description="AI-generated or user-provided title"
    )
    subtitle: str | None = Field(None, description="AI-generated summary/subtitle")


class NoteCreate(BaseModel):
    """Schema for creating a new note."""

    type: NoteType
    content: str | None = Field(None, description="Text content for text notes")


class NoteUpdate(BaseModel):
    """Schema for updating an existing note."""

    content: str | None = Field(None, description="Updated text content")
    title: str | None = Field(None, max_length=255, description="Updated title")
    subtitle: str | None = Field(None, description="Updated subtitle")


class NoteMetadata(BaseModel):
    """AI-generated metadata for a note."""

    title: str = Field(..., max_length=255, description="AI-generated title")
    subtitle: str = Field(..., description="AI-generated brief summary")


class NoteResponse(BaseModel):
    """Schema for note response."""

    id: UUID
    case_id: UUID
    user_id: str
    type: NoteType
    content: str | None = None
    audio_storage_path: str | None = None
    audio_duration_seconds: int | None = None
    audio_mime_type: str | None = None
    title: str | None = None
    subtitle: str | None = None
    is_exported: bool = False
    exported_file_id: UUID | None = None
    created_at: datetime
    updated_at: datetime

    model_config = {
        "from_attributes": True,
    }


class NoteListResponse(BaseModel):
    """Schema for listing notes."""

    notes: list[NoteResponse]
    total: int
    page: int
    per_page: int


class NoteExportRequest(BaseModel):
    """Schema for exporting a note as evidence."""

    description: str | None = Field(
        None, description="Optional description for the exported file"
    )


class NoteExportResponse(BaseModel):
    """Schema for export response."""

    note_id: UUID
    file_id: UUID
    file_name: str
    message: str


class AudioUploadResponse(BaseModel):
    """Schema for audio upload response (before note creation)."""

    storage_path: str
    duration_seconds: int | None = None
    mime_type: str
    size_bytes: int


class GenerateMetadataRequest(BaseModel):
    """Schema for generating AI metadata for a note."""

    note_id: UUID


class GenerateMetadataResponse(BaseModel):
    """Schema for generated metadata response."""

    note_id: UUID
    title: str
    subtitle: str
    content: str | None = None


class AudioDownloadResponse(BaseModel):
    """Schema for audio download URL response."""

    download_url: str
    expires_in: int

File: frontend/src/lib/api/notes.ts
// ABOUTME: API client for notes operations (Sherlock's Diary).
// ABOUTME: Handles note CRUD, audio upload, metadata generation, and evidence export.

import { getToken } from "@/lib/auth-client";

const API_URL = process.env.NEXT_PUBLIC_API_URL || "http://localhost:8080";

export type NoteType = "TEXT" | "AUDIO";

export interface NoteResponse {
  id: string;
  case_id: string;
  user_id: string;
  type: NoteType;
  content: string | null;
  audio_storage_path: string | null;
  audio_duration_seconds: number | null;
  audio_mime_type: string | null;
  title: string | null;
  subtitle: string | null;
  is_exported: boolean;
  exported_file_id: string | null;
  created_at: string;
  updated_at: string;
}

export interface NoteListResponse {
  notes: NoteResponse[];
  total: number;
  page: number;
  per_page: number;
}

export interface NoteCreate {
  type: NoteType;
  content?: string;
}

export interface NoteUpdate {
  content?: string;
  title?: string;
  subtitle?: string;
}

export interface GenerateMetadataResponse {
  note_id: string;
  title: string;
  subtitle: string;
  content?: string | null;
}

export interface NoteExportResponse {
  note_id: string;
  file_id: string;
  file_name: string;
  message: string;
}

export interface AudioDownloadResponse {
  download_url: string;
  expires_in: number;
}

async function getAuthHeaders(): Promise<HeadersInit> {
  const token = await getToken();
  if (!token) {
    throw new Error("Not authenticated");
  }
  return {
    Authorization: `Bearer ${token}`,
  };
}

/**
 * List all notes for a case.
 */
export async function listNotes(
  caseId: string,
  page = 1,
  perPage = 50,
  type?: NoteType,
): Promise<NoteListResponse> {
  const headers = await getAuthHeaders();
  let url = `${API_URL}/api/cases/${caseId}/notes?page=${page}&per_page=${perPage}`;
  if (type) {
    url += `&type=${type}`;
  }
  const res = await fetch(url, { headers });
  if (!res.ok) {
    const error = await res
      .json()
      .catch(() => ({ detail: "Failed to list notes" }));
    throw new Error(error.detail || "Failed to list notes");
  }
  return res.json();
}

/**
 * Get a single note by ID.
 */
export async function getNote(
  caseId: string,
  noteId: string,
): Promise<NoteResponse> {
  const headers = await getAuthHeaders();
  const res = await fetch(`${API_URL}/api/cases/${caseId}/notes/${noteId}`, {
    headers,
  });
  if (!res.ok) {
    const error = await res
      .json()
      .catch(() => ({ detail: "Failed to get note" }));
    throw new Error(error.detail || "Failed to get note");
  }
  return res.json();
}

/**
 * Create a new text note.
 */
export async function createTextNote(
  caseId: string,
  content: string,
): Promise<NoteResponse> {
  const headers = await getAuthHeaders();
  const res = await fetch(`${API_URL}/api/cases/${caseId}/notes`, {
    method: "POST",
    headers: {
      ...headers,
      "Content-Type": "application/json",
    },
    body: JSON.stringify({
      type: "TEXT",
      content,
    }),
  });
  if (!res.ok) {
    const error = await res
      .json()
      .catch(() => ({ detail: "Failed to create note" }));
    throw new Error(error.detail || "Failed to create note");
  }
  return res.json();
}

/**
 * Create a new audio note by uploading an audio file.
 */
export async function createAudioNote(
  caseId: string,
  audioFile: File | Blob,
  durationSeconds?: number,
): Promise<NoteResponse> {
  const headers = await getAuthHeaders();
  const formData = new FormData();
  if (audioFile instanceof File) {
    formData.append("file", audioFile);
  } else {
    // Blob doesn't have a filename, so append one based on type
    let ext = "webm"; // Default for browser recordings
    const type = audioFile.type.toLowerCase();
    if (type.includes("mp4") || type.includes("m4a")) ext = "mp4";
    else if (type.includes("wav")) ext = "wav";
    else if (type.includes("ogg")) ext = "ogg";
    else if (type.includes("mp3")) ext = "mp3";

    formData.append("file", audioFile, `recording.${ext}`);
  }
  if (durationSeconds !== undefined) {
    formData.append("duration_seconds", durationSeconds.toString());
  }
  const res = await fetch(`${API_URL}/api/cases/${caseId}/notes/audio`, {
    method: "POST",
    headers,
    body: formData,
  });
  if (!res.ok) {
    const error = await res
      .json()
      .catch(() => ({ detail: "Failed to upload audio" }));
    throw new Error(error.detail || "Failed to upload audio");
  }
  return res.json();
}

/**
 * Update a note's content or metadata.
 */
export async function updateNote(
  caseId: string,
  noteId: string,
  data: NoteUpdate,
): Promise<NoteResponse> {
  const headers = await getAuthHeaders();
  const res = await fetch(`${API_URL}/api/cases/${caseId}/notes/${noteId}`, {
    method: "PUT",
    headers: {
      ...headers,
      "Content-Type": "application/json",
    },
    body: JSON.stringify(data),
  });
  if (!res.ok) {
    const error = await res
      .json()
      .catch(() => ({ detail: "Failed to update note" }));
    throw new Error(error.detail || "Failed to update note");
  }
  return res.json();
}

/**
 * Delete a note.
 */
export async function deleteNote(
  caseId: string,
  noteId: string,
): Promise<void> {
  const headers = await getAuthHeaders();
  const res = await fetch(`${API_URL}/api/cases/${caseId}/notes/${noteId}`, {
    method: "DELETE",
    headers,
  });
  if (!res.ok) {
    const error = await res
      .json()
      .catch(() => ({ detail: "Failed to delete note" }));
    throw new Error(error.detail || "Failed to delete note");
  }
}

/**
 * Get a signed URL for streaming/downloading audio.
 */
export async function getAudioUrl(
  caseId: string,
  noteId: string,
  inline = true,
): Promise<AudioDownloadResponse> {
  const headers = await getAuthHeaders();
  const res = await fetch(
    `${API_URL}/api/cases/${caseId}/notes/${noteId}/audio?inline=${inline}`,
    { headers },
  );
  if (!res.ok) {
    const error = await res
      .json()
      .catch(() => ({ detail: "Failed to get audio URL" }));
    throw new Error(error.detail || "Failed to get audio URL");
  }
  return res.json();
}

/**
 * Generate AI-powered title and subtitle for a note.
 */
export async function generateMetadata(
  caseId: string,
  noteId: string,
): Promise<GenerateMetadataResponse> {
  const headers = await getAuthHeaders();
  const res = await fetch(
    `${API_URL}/api/cases/${caseId}/notes/${noteId}/generate-metadata`,
    {
      method: "POST",
      headers,
    },
  );
  if (!res.ok) {
    const error = await res
      .json()
      .catch(() => ({ detail: "Failed to generate metadata" }));
    throw new Error(error.detail || "Failed to generate metadata");
  }
  return res.json();
}

/**
 * Export a note as evidence to the case's file library.
 */
export async function exportNoteAsEvidence(
  caseId: string,
  noteId: string,
  description?: string,
): Promise<NoteExportResponse> {
  const headers = await getAuthHeaders();
  const res = await fetch(
    `${API_URL}/api/cases/${caseId}/notes/${noteId}/export`,
    {
      method: "POST",
      headers: {
        ...headers,
        "Content-Type": "application/json",
      },
      body: JSON.stringify({ description }),
    },
  );
  if (!res.ok) {
    const error = await res
      .json()
      .catch(() => ({ detail: "Failed to export note" }));
    throw new Error(error.detail || "Failed to export note");
  }
  return res.json();
}
